crates/iota-indexer/src/apis/write_api.rs

```diff
@@ -2,30 +2,440 @@
+        checkpoint_handler::{CheckpointHandler, try_extract_df_kind},
+    },
+    indexer_reader::IndexerReader,
+    metrics::IndexerMetrics,
+    models::{
+        display::StoredDisplay,
+        event_indices::OptimisticEventIndices,
+        events::StoredEvent,
+        transactions::{StoredTransaction, TxGlobalOrder},
+        tx_indices::OptimisticTxIndices,
+    },
+    run_query_async,
+    schema::tx_global_order,
+    spawn_read_only_blocking,
+    store::{IndexerStore, PgIndexerStore},
+    transactional_blocking_with_retry,
+    types::{
+        EventIndex, IndexedDeletedObject, IndexedEvent, IndexedObject, IndexedTransaction,
+        IndexerResult, IotaTransactionBlockResponseWithOptions, TxIndex,
+    },
+};
 
 pub(crate) struct WriteApi {
     fullnode: HttpClient,
+    fullnode_rest_client: iota_rest_api::Client,
+    inner: IndexerReader,
+    store: PgIndexerStore,
+    metrics: IndexerMetrics,
 }
 
 impl WriteApi {
-    pub fn new(fullnode_client: HttpClient) -> Self {
+    pub fn new(
+        fullnode_client: HttpClient,
+        fullnode_rest_client: iota_rest_api::Client,
+        inner: IndexerReader,
+        store: PgIndexerStore,
+        metrics: IndexerMetrics,
+    ) -> Self {
         Self {
             fullnode: fullnode_client,
+            fullnode_rest_client,
+            inner,
+            store,
+            metrics,
+        }
+    }
+
+    async fn execute_and_optimistically_index_tx_effects(
+        &self,
+        tx_bytes: Base64,
+        signatures: Vec<Base64>,
+        options: Option<IotaTransactionBlockResponseOptions>,
+    ) -> RpcResult<IotaTransactionBlockResponse> {
+        let tx_data: TransactionData =
+            bcs::from_bytes(&tx_bytes.to_vec().map_err(IndexerError::FastCrypto)?)
+                .map_err(IndexerError::Bcs)?;
+        let mut sigs = Vec::new();
+        for sig in signatures {
+            sigs.push(
+                GenericSignature::from_bytes(&sig.to_vec().map_err(IndexerError::FastCrypto)?)
+                    .map_err(IndexerError::FastCrypto)?,
+            );
+        }
+        let transaction = Transaction::from_generic_sig_data(tx_data, sigs);
+
+        // TODO: shouldn't return type below be from rust-sdk types?
+        // Is this type correct?
+        let result = self
+            .fullnode_rest_client
+            .execute_transaction(
+                &ExecuteTransactionQueryParameters {
+                    events: true,
+                    balance_changes: false,
+                    input_objects: true,
+                    output_objects: true,
+                },
+                &transaction,
+            )
+            .await
+            .map_err(|e| IndexerError::Generic(e.to_string()))?;
+
+        let TransactionExecutionResponse {
+            effects,
+            finality: _,
+            events,
+            balance_changes: _,
+            input_objects,
+            output_objects,
+        } = result;
```

Comment:
You can [ignore remaining parts](https://doc.rust-lang.org/book/ch19-03-pattern-syntax.html#remaining-parts-of-a-value-with-)
```suggestion
        let TransactionExecutionResponse {
            effects,
            events,
            input_objects,
            output_objects,
            ..
        } = result;
```

But I think you can even spare the reader of the additional lines of code if you just access the `result` fields directly later (e.g. `result.input_objects`).

Reply:
I think it looks clearer if this struct is unpacked. I will go with `..` for now.